### ISSUE TITLE
Potential New Combinators for Lens-Errors

### ISSUE THREAD

**Issue by rohit507** [ChrisPenner/lens-errors#2]
So I've being playing with my code, and I've got a few more combinators for you to consider: 

In `lens` there's a number of combinators that allow mere mortals to create 
optics: `lens`,`iso`,`prism`, etc... 

I was able to come up with one if you add a Monad constraint to `LensFail`.  All of the instances of `lensFail` seem to also have Monad instances, so it might 
just do to add that requirement to `LensFail`. 

```
lensErr :: (forall f. (Monad f, LensFail e f) => s -> f a)
        -> (forall f. (Monad f, LensFail e f) => s -> b -> f t )
        -> (forall f. (Monad f, LensFail e f) => LensLike f s t a b)
lensErr examine modify a2fb s = examine s >>= a2fb >>= modify s
```

Also like Lens, a more concrete type synonym for the error: 
```
type LensErr  e s t a b = forall f. (Monad f, LensFail e f) => LensLike f s t a b
type LensErr' e s   a   = LensErr e s s a a
```

I also happen to have cases where I want to compose a bunch of autogenerated prisms in parallel, and end up doing something like: 

```
data Lit
  = FloatLit Double
  | IntLit Integer
  | BoolLit Bool
  | TextLit Text
  | SymbolLit Text
  | UIDLit
  deriving (Eq, Ord, Generic, Show, NFData)

libLiteralL :: Lens' P.ValueLit Lit
libLiteralL = lens (getter . fromMaybe err . view P.maybe'type') (flip fsetter)
  where
    getter :: P.ValueLit'Type -> Lit
    getter (P.ValueLit'Floating l) = FloatLit  (l ^. P.val)
    getter (P.ValueLit'Integer  l) = IntLit    (toInteger $ l ^. P.val)
    getter (P.ValueLit'Boolean  l) = BoolLit   (l ^. P.val)
    getter (P.ValueLit'Text     l) = TextLit   (l ^. P.val)
    getter (P.ValueLit'Symbol   l) = SymbolLit (l ^. P.val)
    getter (P.ValueLit'Uid      _) = UIDLit

    fsetter :: Lit -> P.ValueLit -> P.ValueLit
    fsetter (FloatLit  f) = set (P.floating . P.val) f
    fsetter (IntLit    f) = set (P.integer  . P.val) (fromInteger f)
    fsetter (BoolLit   f) = set (P.boolean  . P.val) f
    fsetter (TextLit   f) = set (P.text     . P.val) f
    fsetter (SymbolLit f) = set (P.symbol   . P.val) f
    fsetter UIDLit        = set P.uid defMessage

    err = error "Expression is not a literal."  
```

if you look at fsetter, you can see how there's a pile of  generated prisms `P.floating`, `P.integer` that allow me to distinguish between the various constructors of `P.ValueLit`. Likewise, prisms for `Lit` are also easily autogenerated. 

If there's some sort of parallel composition operator, like what follows, then I could use a nicer style: 
```
-- | Try the first LensErr and if that fails, then the second.
--   Basically, composition via alternative for LensErr.
--
--   I don't think this makes a legal lens unless the lens used is completely
--   determined by s.
tryFizzle :: LensErr e s t a b -> LensErr e s t a b -> LensErr e s t a b
tryFizzle = undefined

-- | LensErr that always fizzles with some error.
alwaysFizzle :: (s -> e) -> (s -> b -> e) -> LensErr e s t a b
alwaysFizzle = undefined
```

Like this, where anything prefixed by an underscore (like `_FloatLit`) is a prism into a `Lit`:

```
libLiteral :: LensErr' e P.ValueLit Lit 
libLiteral = (P.floating . P.val . _FloatLit)
    `tryFizzle` (P.integer . P.val . _IntLit)
    `tryFizzle` (P.boolean . P.val . _BoolLit)
    ... 
     `tryFizzle` alwaysFizzle (InvalidValueLit) (InvalidLit) 
```

**Comment by ChrisPenner** [comment]
A few thoughts;

> lensErr :: (forall f. (Monad f, LensFail e f) => s -> f a)
        -> (forall f. (Monad f, LensFail e f) => s -> b -> f t )
        -> (forall f. (Monad f, LensFail e f) => LensLike f s t a b)

This is basically a Traversal, but actually is more restrictive (this isn't a valid lens OR traversal due to the Monad constraint). You can make your own lens-fails using the Traversable signature (`(a -> f b) -> s -> f t`); and then just use `fizzle :: e -> f a` inside wherever you like 😄; Perhaps I should add a guide or document it, but it should roughly line up with the custom traversal guide [here](https://lens-by-example.chrispenner.ca/articles/traversals/writing-traversals). Also note that adding a Monad constraint will prevent it from working with the error-getters like `^&.` and `^&..` since those use `Const` which doesn't have a Monad instance.

> type LensErr  e s t a b = forall f. (Monad f, LensFail e f) => LensLike f s t a b
> type LensErr' e s   a   = LensErr e s s a a

This is a good idea; I'll add something like this (with the more permissive `Applicative` constraint) soon, but I'm actually moving to a different continent tomorrow so it might take a few days 😋 

```
libLiteral :: LensErr' e P.ValueLit Lit 
libLiteral = (P.floating . P.val . _FloatLit)
    `tryFizzle` (P.integer . P.val . _IntLit)
    `tryFizzle` (P.boolean . P.val . _BoolLit)
    ... 
     `tryFizzle` alwaysFizzle (InvalidValueLit) (InvalidLit) 
```

You can do this using `failing`; unfortunately it can't take a LensErr on the left-hand side, but it can on the right-hand side; so I think it'll work for this case; e.g.

```
libLiteral :: LensErr' e P.ValueLit Lit 
libLiteral = (P.floating . P.val . _FloatLit)
    `failing` (P.integer . P.val . _IntLit)
    `failing` (P.boolean . P.val . _BoolLit)
    ... 
     `failing` fizzleWith (const InvalidValueLit) 
```

^ That should work for this case; it's sad that `failing` can't take a LensErr on the left, but I'm not even sure exactly what that would mean, and would probably stack errors in a weird way.

I don't really understand how `alwaysFizzle` is meant to work:

```
alwaysFizzle :: (s -> e) -> (s -> b -> e) -> LensErr e s t a b
```
The signature suggests it has access to both an `s` and `b`, and that it combines them somehow, but the `s -> e` seems to be redundant in light of `s -> b -> e`;  is `fizzleWith` not sufficient for your needs?

**Comment by rohit507** [comment]
> This is basically a Traversal, but actually is more restrictive (this isn't a valid lens OR traversal due to the Monad constraint). You can make your own lens-fails using the Traversable signature (`(a -> f b) -> s -> f t`); and then just use `fizzle :: e -> f a` inside wherever you like ; Perhaps I should add a guide or document it, but it should roughly line up with the custom traversal guide [here](https://lens-by-example.chrispenner.ca/articles/traversals/writing-traversals). Also note that adding a Monad constraint will prevent it from working with the error-getters like `^&.` and `^&..` since those use `Const` which doesn't have a Monad instance.

Good to know, giving this a try now. 

> > type LensErr  e s t a b = forall f. (Monad f, LensFail e f) => LensLike f s t a b
> > type LensErr' e s   a   = LensErr e s s a a
> 
> This is a good idea; I'll add something like this (with the more permissive `Applicative` constraint) soon, but I'm actually moving to a different continent tomorrow so it might take a few days 

Cool, good luck with the move. 

> I don't really understand how `alwaysFizzle` is meant to work:
> 
> ```
> alwaysFizzle :: (s -> e) -> (s -> b -> e) -> LensErr e s t a b
> ```
> 
> The signature suggests it has access to both an `s` and `b`, and that it combines them somehow, but the `s -> e` seems to be redundant in light of `s -> b -> e`; is `fizzleWith` not sufficient for your needs?

The idea is you can get different errors when trying to extract something from s or when trying to insert something into s. My project requires extracting a sub-language from a larger language. So there are different errors when the input is badly formed or you're trying to extract something that isn't there (`s -> e`) compared to when your attempted modification is itself malformed (`s -> b -> e`). 

I'm not confident that this is actually a reasonable notion, so tell me if I'm asking for something lenses weren't built for.

**Comment by rohit507** [comment]
Trying an alternate way to construct a lensErr makes it pretty obvious that you can't have errors for setter operations without a monad constraint: 

```
lensErr :: (s -> Either e a) -> (s -> b -> Either e t) -> LensErr e s t a b
lensErr sea sbt afb s = case sea s of
  Left e -> fizzle e
  Right a -> _
```

I'm not really sure whether what I'd like is even doable at this point, short of adding a vacuous monad instance to const. :/ 

At the very least, here's a function that allows for easy construction of error lenses, in the same style as `iso`, `prism`, and `lens`:

```
lensErr :: (s -> Either e a) -> (s -> b -> t) -> LensErr e s t a b
lensErr sea sbt afb s = case sea s of
  Left e -> fizzle e
  Right a -> sbt s <$> afb a
```

At the very least, having that sort of lens constructor in the library makes it easier for people who don't really understand the inner working of lenses to work with `lens-errors`.

**Comment by rohit507** [comment]
I tried to play around with this on my own fork https://github.com/rohit507/lens-error and creating vacuous monad instances seems to work. 

In particular I altered `LensFail` a bit 
```

class LensFail e f | f -> e where
  [...]
  fizzJoin :: f (f a) -> f a

instance Monoid a => LensFail e (Const (e, a)) where
  fizzJoin (Const t) = Const t

instance LensFail e (Const (Either e a)) where
  fizzJoin (Const t) = Const t

instance LensFail e (Const (Validation e a)) where
  fizzJoin (Const t) = Const t

instance LensFail e (Either e) where
  fizzJoin = join

instance LensFail e (Validation e) where
  fizzJoin (Failure e) = Failure e
  fizzJoin (Success a) = a
```

Which lets us define the combinator I want: 

```
lensErr :: forall f e s t a b.  (Applicative f, LensFail e f)
        => (s -> Either e a) -> (s -> b -> Either e t) -> LensLike f s t a b
lensErr sea sbet afb s = case sea s of
  Left e -> fizzle e
  Right a -> fizzJoin (either fizzle pure . sbet s <$> afb a)
```

This lets me differentiate between access errors and modification errors.

```
foo :: LensError' String (a,Int) Int
foo = lensErr (\ (_, n)   -> if n < 0 then Left "Get Err" else Right n)
              (\ (s, _) n -> if n < 0 then Left "Set Err" else Right (s,n))

main :: IO ()
main = do
  print $ ("bar", 10) ^&? foo       -- > Success 10
  print $ ("bar", -3) ^&? foo       -- > Failure "Get Err"
  print $ ("bar", -3) & foo .&~ 3   -- > Failure "Get Err"
  print $ ("bar", 3) & foo .&~ 3    -- > Success ("Bar", 3)  
  print $ ("bar", 3) & foo .&~ (-3) -- > Failure "Set Err"
```

As a bonus, I didn't have to modify any of the combinators you define, just allow an alternate way to construct a lens.

**Comment by ChrisPenner** [comment]
@rohit507 Ahh; interesting. I see what you're motivated by now, but not quite convinced it's general enough to include `lensErr` in the lib;

> I'm not really sure whether what I'd like is even doable at this point, short of adding a vacuous monad instance to const. :/

Note that the `Vacuous` monad constraint you're talking about is typically encoded in Lens as `Functor f, Contravariant f` which necessitates that the `a` in `f a` is a phantom parameter.

E.g. you can implement fizzJoin for `Const` like this, using [phantom](https://hackage.haskell.org/package/base/docs/Data-Functor-Contravariant.html#v:phantom):

```haskell
  fizzJoin :: (Contraveriant f, Functor f) => f (f a) -> f b
  fizzJoin = phantom
```

Because: `phantom :: (Functor f, Contravariant f) => f a -> f b`.

If you generally want different errors based on failure in modification or in access I think you can get most of the way there using the existing combinators, so you can build a normal lens chain with a failure built into 'access', which will trigger itself automatically, but then run modifications with `%%&~ :: LensLike (Validation e) s t a b -> (a -> Validation e b) -> s -> Validation e t `  which will let you fail at modification time; and you can add additional context to those errors by adding an `adjustingErrorsWith :: LensFail e f => (s -> e -> e) -> LensLike' f s s` in the chain; but I definitely understand that this is a bit more awkward.

Regardless, I think a version of `lensErr` would be helpful to expose; but I'd want to do it without the Monad'ish constraint if possible. 

I dunno though, maybe I'm too picky on that one haha. I'll have to play around a bit 😬 

Thanks for using this for real use-cases; definitely appreciate the feedback, hope it doesn't seem like I'm shutting things down, just trying to distill and generalize your needs a bit from your concrete use-cases.

**Comment by rohit507** [comment]
No worries, it's straightforward enough to wrap things up how I want them anyway. 

Really, I need to:
  - Enforce invariants over the structure I'm working with.
  - Enforce that updates to a structure follow rules that are stricter than just observing those invariants in the output.
  - Have good error messages 

Which is frankly an outcome of a really weird use case where we have a bunch of systems that work with the same large structures (`s`) but each can only manipulate "views" (`a`, `b`, etc.. such that all those types are factors of `s` but there is no `c` such that `s = a * b * c` ).  

I don't really expect the same set of restrictions to pop up often for others.

**Comment by ChrisPenner** [comment]
@rohit507 I changed my mind after digging into it a bit; all the arguments you made make a good amount of sense 👍 

New version published [here](http://hackage.haskell.org/package/lens-errors) 😄 

It adds the `Fizzler` type aliases and `fizzler` construction helper which should hopefully cover your cases; let me know if they don't 😃
